document/JS.bs

```diff
@@ -0,0 +1,917 @@
+
+Issue: In future versions, WebAssembly may
+be loaded and run directly from an HTML `<script type='module'>` tag—and
+any other Web API that loads ES6 modules via URL—as part of
+[ES6 Module integration](https://github.com/WebAssembly/design/issues/1087).)
+
+Note: WebAssembly JS API declaration file for TypeScript can be found [here](https://github.com/01alchemist/webassembly-types/blob/master/webassembly.d.ts) which enable autocompletion and make TypeScript compiler happy.
+
+<h2 id="sample">Sample API usage</h2>
+
+<p><em>This section is non-normative.</em></p>
+
+Given `demo.wat` (encoded to `demo.wasm`):
+
+```lisp
+(module
+    (import "js" "import1" (func $i1))
+    (import "js" "import2" (func $i2))
+    (func $main (call $i1))
+    (start $main)
+    (func (export "f") (call $i2))
+)
+```
+
+and the following JavaScript, run in a browser:
+
+```javascript
+var importObj = {js: {
+    import1: () => console.log("hello,"),
+    import2: () => console.log("world!")
+}};
+fetch('demo.wasm').then(response =>
+    response.arrayBuffer()
+).then(buffer =>
+    WebAssembly.instantiate(buffer, importObj)
+).then(({module, instance}) =>
+    instance.exports.f()
+);
+```
+
+<h2 id="webassembly-namespace">The WebAssembly Namespace</h2>
+
+<pre class="idl">
+dictionary WebAssemblyInstantiatedSource {
+    required Module module;
+    required Instance instance;
+};
+
+[Exposed=(Window,Worker,Worklet)]
+namespace WebAssembly {
+    boolean validate(BufferSource bytes);
+    Promise&lt;Module> compile(BufferSource bytes);
+
+    Promise&lt;WebAssemblyInstantiatedSource> instantiate(
+        BufferSource bytes, optional object importObject);
+
+    Promise&lt;Instance> instantiate(
+        Module moduleObject, optional object importObject);
+};
+</pre>
+
+<!--
+Should we include notes describing what the functions do, as the HTML spec does? It could look like this:
+
+Note:
+  WebAssembly.validate(|bytes|) synchronously validates bytes of WebAssembly, returning true if the validation was successful.
+  WebAssembly.compile(|bytes|) asynchronously validates and complies bytes of WebAssembly into a Module.
+  WebAssembly.instantiate(|bytes|, |importObject|) asynchronously compiles and instantiates a WebAssembly module from bytes of source.
+  The WebAssembly.instantiate(|moduleObject|, |importObject|) asynchronously instantiates a compiled module.
+-->
+
+<div algorithm>
+  To <dfn>compile a WebAssembly module</dfn> from a {{BufferSource}} |bytes|, perform the following steps:
+    1. Let |module| be [=decode_module=](|bytes|). If |module| is [=error=], return [=error=].
+    1. If [=validate_module=](|module|) is [=error=], return [=error=].
+    1. Return |module|.
+
+    Note: The custom [=name section=] is not validated here; an invalid name section does not impact behavior.
```

Comment:
Can we just drop this note? It is both redundant and irrelevant to the JS API specifically. There isn't even a defined notion of "validation" for the name section.

Reply:
I'm confused how we could drop the note because later in the spec we UTF-8 decode the names and assert that this will not fail because they've been validated. Is that not true?

Reply:
The JS API does not use, expose or even mention the name section, it is completely impartial to its existence. The name section is solely a debugging aid for pretty-printing Wasm code, which is outside the scope of this spec.

Reply:
Moreover, it is explicitly allowed to be malformed, in which case engines are supposed to ignore it.

Reply:
I guess I assumed module_exports() / module_imports() / custom sections related to the name section. (See https://littledan.github.io/spec/document/JS.html#instantiate-a-webassembly-module step 16.1/2, https://littledan.github.io/spec/document/JS.html#dom-module-exports-moduleobject step 3.1/2, https://littledan.github.io/spec/document/JS.html#dom-module-imports-moduleobject step 3.1/2/3, and https://littledan.github.io/spec/document/JS.html#dom-module-customsections-moduleobject-sectionname step 3.1/2.)

Is that not correct? In that case, the "because" clause in each of those asserts probably needs updating.

Reply:
Externally visible import/export names are part of the main binary format, and required to be checked during decoding. The name section isn't, it merely associates print names with local indices, none of which are accessible externally.

Reply:
OK, so "`name` of a custom section" is different from the "custom name section"?

Reply:
Yes, every custom section has a name identifying its format. There is one particular name, namely `"name"`, which identifies the name section. :)

So the `customSections` function could return the name section, among others, but only as a completely uninterpreted sequence of raw bytes.